consensus.go

```diff
@@ -600,25 +614,25 @@ func (p *Pbft) runRoundChangeState(ctx context.Context) {
 
 // --- communication wrappers ---
 
-func (p *Pbft) sendRoundChange() {
-	p.gossip(MessageReq_RoundChange)
+func (p *Pbft) sendRoundChange() error {
+	return p.gossip(MessageReq_RoundChange)
 }
 
-func (p *Pbft) sendPreprepareMsg() {
-	p.gossip(MessageReq_Preprepare)
+func (p *Pbft) sendPreprepareMsg() error {
+	return p.gossip(MessageReq_Preprepare)
 }
 
-func (p *Pbft) sendPrepareMsg() {
-	p.gossip(MessageReq_Prepare)
+func (p *Pbft) sendPrepareMsg() error {
+	return p.gossip(MessageReq_Prepare)
 }
 
-func (p *Pbft) sendCommitMsg() {
-	p.gossip(MessageReq_Commit)
+func (p *Pbft) sendCommitMsg() error {
+	return p.gossip(MessageReq_Commit)
 }
 
-func (p *Pbft) gossip(typ MsgType) {
+func (p *Pbft) gossip(msgType MsgType) error {
```

Comment:
So what should we do if there is a gossip error? What type of errors can we have in the gossip protocol? My understanding is that the Transport and Gossip function are async functions that should not fail to the user but handle any errors themselves.

Internally, a gossip transports enqueues all the possible messages from the Peer and does its best to broadcast them. I think we could even remove the error from the function altogether. Similar though applies to Sign. Thus, I think we can remove this errors from the gossip functions.

Reply:
sure, I will make adjustments.

Reply:
sure, I will make adjustments.

Reply:
> So what should we do if there is a gossip error? 

We had this example on our mind when been thinking through it and thought maybe it is better to return an `error` object and caller can decide how to handle it. The question in the snippet below is: is it ok here to proceed with sending `prepare` messages if for some reason `sendPreprepareMsg` failed? If the answer is affirmative, then we should discuss and agree upon error handling strategy for each use case which can happen.
```
func (p *Pbft) runAcceptState(ctx context.Context) {
...
   // send the preprepare message
   p.sendPreprepareMsg()
   // send the prepare message since we are ready to move the state
   p.sendPrepareMsg()
...
}
```
On the other hand, we don't mind with only logging out errors, but thought that returning error object would be convenient.

> What type of errors can we have in the gossip protocol?

We don't know that yet, those are implementation details of validator seal signing (`Sign` function), gossiping on transport layer (`Gossip` function) etc. Namely each function which is being invoked within a 
`func (p *Pbft) gossip(msgType MsgType)` function.

Reply:
What are you going to do in the PBFT state machine if sending Preprepare fails?

Reply:
I was thinking maybe interrupting state machine loop and propagate an error back to the client immediately and it could decide about some retry logic, if it is important to send `prepare` message only after successful `pre-prepare` message sending. 
However, I got your point as well and seems this is not the case:
>  ...a gossip transports enqueues all the possible messages from the Peer and does its best to broadcast them.

which suggests that a node is doing its best to deliver all the messages to the peers and there is anyway `RoundChange` state when some essential condition is unsatisfied.

Just for the note, we have already reverted this change, just wanted to discuss around it a little bit more. :slightly_smiling_face: